### DISCUSSION TITLE
Include configurations from env variables?

### DISCUSSION THREAD

**Discussion by ennerf** [hydraulic-software/conveyor#99] [Ideas]
Is there a way to include configurations from environment variables? Instead of creating a [ci.conveyor.conf](https://github.com/HebiRobotics/QuickBuffers/blob/7941ddac96ebf1a23c554633c0af33a7bb736e27/conveyor-ci.conf) wrapper that needs [lots of environment variables](https://github.com/HebiRobotics/QuickBuffers/blob/7941ddac96ebf1a23c554633c0af33a7bb736e27/.github/workflows/release.yml#L178-L185), e.g.,

```hocon
// ci.conveyor.conf
include required("conveyor.conf")

app {
  sign = true;
  signing-key = ${env.SIGNING_KEY}

  windows {
    signing-key {
      digi-cert-one {
        api-key = ${env.DIGICERT_API_KEY}
        auth-certificate = ${env.DIGICERT_AUTH_CERT}
        password = ${env.DIGICERT_AUTH_PW}
      }
    }
    signing-key-alias = ${env.DIGICERT_SIGNING_KEY_ALIAS}
  }

  mac { 
    certificate = "deploy/apple.cer"
    notarization {
      apple-id = ${env.APPLE_ID}
      team-id = ${env.APPLE_TEAM_ID}
      app-specific-password = ${env.APPLE_APP_PASSWORD}
    }
  }
}
```

I'd prefer to include some environment variable in JSON/HOCON format that contains the shared configurations, e.g.,

```
// conveyor.conf
include json(${ env.CONVEYOR_DEFAULT_RELEASE_CONFIG })
include json(${ env.CONVEYOR_WINDOWS_SIGNING_CONFIG })
```

Empty or undefined variables should do nothing. That'd make dealing with CI a lot easier. I'd prefer to not write the env variable to a `default-config` file that is potentially overwritten or not cleaned up properly (on self-hosted runners).

**Comment by mikehearn** [comment]
I think the only way to do this currently is to use hashbang includes:

```
include required("#! cat $MY_ENV_VAR")
```

Allowing interpolation into include paths is interesting and something we've poked at before for other reasons, but it's easy to end up with confusing circularity issues (e.g. what happens if the included file redefines the variabled used as part of the include path).

**Reply by ennerf** [reply]
Rather than a path interpolation I was thinking of an env variable that contains the JSON text

**Reply by mikehearn** [reply]
So I guess "echo $ENV" should work?
